Optimization of workflow

qiyunzhu opened the pull request:
This thread involves several optimizations aiming at reducing runtime which is current hours to days on large datasets (e.g., some hundreds of GB of alignment files), as well as bug fixes and trivial modifications. Currently, the changes are:

- [ ] CIGAR string parsing. Commit [64b8b69](https://github.com/qiyunzhu/woltka/pull/39/commits/64b8b69e9c0522ac8d7ee98c409f1f951dd51b10).
  - See issue #38 for details.
- [ ] Count list. Code is not changed but benchmarks are provided.
  - See added comment to issue #9 by @gwarmstrong.
- [ ] Avoided feature `None`. Commit [4635406](https://github.com/qiyunzhu/woltka/pull/39/commits/46354062e2f4a4c1a6af7d6b51e0e641d9be3765).
  - This bug partially caused the problem reported by @adswafford in issue #34. By default, Woltka allows sequences to be unclassified -- and they are assigned to taxon `None`. However `None` cannot be sorted together with string taxa. Therefore, this patch removes `None` from the resulting profiles.
- [ ] Allowed multiple `--names` files. Commit [ceecf64](https://github.com/qiyunzhu/woltka/pull/39/commits/ceecf64ac4a25dfabef80702f26f28dbdfcbfa9e).
  - A common use case is in functional classification. Each hierarchy (e.g., gene, module, pathway) has its own names file. Allowing them to be supplied simultaneously is favorable.
- [ ] Renamed `ranktb` as `rank_table`.
  - Also provided rationales in issue #16.
- [ ] Screen output enrichment. Commit [1ab7281](https://github.com/qiyunzhu/woltka/pull/39/commits/1ab728198bbe024cabb1b4615a1ccc4855833d2d).

qiyunzhu commented:
I am going to stop here with a reasonable-sized pull request. @gwarmstrong @fedarko @emikoifish This should already addressed multiple (but not all) questions proposed by @adswafford @swandro and @gregpoore.

emikoifish commented on woltka/cli.py:
```diff
@@ -99,19 +99,19 @@ def gotu(ctx, **kwargs):
     '--map', '-m', 'map_fps', type=click.Path(exists=True), multiple=True,
-    help=('Map(s) of subjects or lower classification units to higher ones. '
-          'Can accept multiple maps.'))
+    help=('Map of ower classification units to higher ones. Can accept '
```
typo, think it should be lower

qiyunzhu commented on woltka/cli.py:
```diff
@@ -99,19 +99,19 @@ def gotu(ctx, **kwargs):
     help='Map of lineage strings. Can accept Greengenes-style rank prefix.')
 @click.option(
-    '--rank-table', 'ranktb_fp', type=click.Path(exists=True),
+    '--rank-table', 'ranktbl_fp', type=click.Path(exists=True),
```
Good point! There was no specific reason except for that I wanted to keep lines short (see workflow.py: line 50). If you think `--rank-table` is preferred, I am happy to make that change. What do you think?

emikoifish replied:
I would prefer `rank-table_fp` because it would make it consistent with the other naming convention.

qiyunzhu replied:
Sounds good! Modified.

qiyunzhu commented on woltka/workflow.py:
```diff
@@ -754,8 +755,9 @@ def write_profiles(data:        dict,
         # multiple output files
         makedirs(fp, exist_ok=True)
-        rank2fp = {x: join(fp, f'{x}.biom') for x in ranks}
         is_biom = is_biom is not False
```
Thanks for noting this! This `is_biom` can actually have three values: `None`, `True` and `False`. This is the interesting part of the Click command-line flags. I added some description of the logic in the docstring of this function.

qiyunzhu commented on woltka/ordinal.py:
```diff
@@ -215,7 +215,7 @@ def whether_prefix(coords):
     -------
     bool
-        Whether gene Ids should be prefixed.
+        Whether gene IDs should be prefixed.
```
Good catch! I have changed multiple instances through the code to make this pattern consistent: for sentences with first word capitalized and large character a symbol (e.g., period), I did "ID"; for in-line comments starting with "#", I did "Id".

qiyunzhu commented:
Hi @emikoifish I addressed your comments. Would you like to take another look?

Hi @gwarmstrong and @fedarko Would you be interested in adding some comments?
